stlab/forest_algorithms.hpp

```diff
@@ -56,14 +58,14 @@ struct forest_insert_iterator {
 };
 
 template <class Forest>
-auto forest_inserter(Forest& f) {
-    return forest_insert_iterator(f);
+auto make_inserter(Forest& f) {
+    return insert_iterator(f);
 }
 
 /**************************************************************************************************/
 
 template <class I, class O, class P, class UP>
-auto transform_forest(I first, I last, O out, P&& proj, UP&& pred) {
+auto transform(I first, I last, O out, P proj, UP pred) {
     for (; first != last; ++first) {
         ++out;
```

Comment:
This is skipping the first output? What is this function doing? `O` is not an output iterator but another forest? I think just `transform` is likely a bad name.

Reply:
I called it `transform` here because I thought in the context of forests it's separate enough that the name would be usable. I once had it written as `transcribe` (and use that name elsewhere) but am open to any other names that might be better.

The point of the algorithm is to transform the forest nodes `[first, last)` by some projection function, and insert the projections into another forest represented by the output iterator `out`. The big difference between this algorithm and the standard algorithms is that the shape of the resulting forest will exactly match the shape of the source. The resulting forest starts out completely empty, and when this algorithm is done its nodes will have the same shape as the original forest.

Reply:
This algorithm generalizes the traversal patterns needed to replicate the source forest shape into the destination. It's used in two places in this file, which is why it is it's own routine. It really needs to be used in combination with `forests::inserter`. I'm not super excited that they're separate entitiees; maybe they just need to be a single algorithm, no `inserter`?